### PR TITLE
#49 - discount field added and its calculations

### PR DESCRIPTION
**Pull request by PragatiDoshi-BTC**

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def total_amount
     total = line_items.inject(0.0) do |total,line_item|
       total + line_item.line_total.to_f
     end
-    total
+    total.to_f - (total.to_f*discount.to_f)/100.00
```
Comment (amitpatelx) [inline app/models/invoice.rb:19]: required space surrounding operators

File: app/views/invoices/_form.html.haml [hunk 2]
```diff
@@ -64,6 +65,13 @@
         .box-body
           .form-group
             - if @invoice.new_record?
+              = label_tag :discount,"Invoice Discount ( % ): "
+              = text_field_tag "invoice[discount]", "0.00"
+            - else
+              = label_tag :discount,"Invoice Discount ( % ): "
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: Label is common in both branches

File: app/views/invoices/_form.html.haml [hunk 3]
```diff
@@ -64,6 +65,13 @@
...
+              = text_field_tag "invoice[discount]", @invoice.discount.to_s
+          .form-group
+            - if @invoice.new_record?
               = label_tag "","Invoice Total (AFN) : ", { id: "show_currency" }
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: Remove `(AFN)`

File: app/views/invoices/_form.html.haml [hunk 4]
```diff
@@ -86,7 +94,20 @@
             = f.button "Save & Send by Email", class: "btn btn-primary", id: "send_button", data: { disable_with: "Saving..." }
 
 :javascript
+  $(document).ready(function() {
+    $(function() {
+      $('.datepicker').datepicker();
+    });
+    $("#invoice_total").on('change', function() {
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: move invoice relate js to `invoice.js`

**Comment by amitpatelx** [comment]
@PragatiDoshi-BTC @ParthivPatel-BTC it requires lot of refactoring in JS. As of now you can merge after fixing CR comments.

I have to look after the JS refactoring

File: app/views/invoices/_form.html.haml [hunk 5]
```diff
@@ -62,13 +63,27 @@
     .col-md-5
       .box.remove_border_element
         .box-body
+          .form-group{:style => "margin-left:25px;"}
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: Use ruby 2 style hash notation

File: app/views/invoices/_form.html.haml [hunk 6]
```diff
@@ -62,13 +63,27 @@
...
+            - if @invoice.new_record?
+              = label_tag :line_total,"Line Total: "
+              = text_field_tag "line_total", "0.00"
+            - else
+              = label_tag :line_total,"Line Total: "
+              = text_field_tag "line_total", number_with_precision(@invoice.total_amount+((@invoice.total_amount.to_f*@invoice.discount.to_f)/100.00), :precision => 2)
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: Move calculation to a helper method

File: app/views/invoices/_form.html.haml [hunk 7]
```diff
@@ -62,13 +63,27 @@
...
           .form-group
             - if @invoice.new_record?
-              = label_tag "","Invoice Total (AFN) : ", { id: "show_currency" }
+              = label_tag :discount,"Discount ( % ): "
+              = text_field_tag "invoice[discount]", "0.00"
+            - else
+              = label_tag :discount,"Discount ( % ): "
+              = text_field_tag "invoice[discount]", number_with_precision(@invoice.discount, :precision => 2)
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: Use ruby 2 style hash notation

File: app/views/invoices/_form.html.haml [hunk 8]
```diff
@@ -62,13 +63,27 @@
...
+          .form-group{:style => "margin-left:15px;"}
+            - if @invoice.new_record?
+              = label_tag "","Total (AFN) : ", { id: "show_currency" }
               = text_field_tag :invoice_total, "0.00"
             - else
-              = label_tag "","Invoice Total (#{@invoice.currency_code.to_s.upcase}) : ", { id: "show_currency" }
-              = text_field_tag :invoice_total, @invoice.total_amount.to_s
+              = label_tag "","Total (#{@invoice.currency_code.to_s.upcase}) : ", { id: "show_currency" }
+              = text_field_tag :invoice_total, number_with_precision(@invoice.total_amount, :precision => 2)
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml:86]: Use ruby 2 style hash notation

File: app/views/invoices/_form.html.haml [hunk 9]
```diff
@@ -82,11 +97,27 @@
...
             = hidden_field_tag :email_send, :value => "false", :id => "email_send"
-            = f.submit "Save as Draft", class: "btn btn-primary", id: "submit_button", data: { disable_with: "Saving..." }
+            = f.button "Save as Draft", class: "btn btn-primary", id: "submit_button", data: { disable_with: "Saving..." }
             = f.button "Save & Send by Email", class: "btn btn-primary", id: "send_button", data: { disable_with: "Saving..." }
 
 :javascript
+  $(document).ready(function() {
+
+    $('#invoice_date_of_issue').datepicker({autoclose: true});
+    $('#invoice_paid_to_date').datepicker({autoclose: true});
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: you can combine binding of both like `$('#invoice_date_of_issue, #invoice_paid_to_date').datepicker({autoclose: true});` in a single expression.

File: app/views/invoices/_form.html.haml [hunk 10]
```diff
@@ -82,11 +97,27 @@
...
+
+    $("#invoice_total").on('change', function() {
+
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: remove extra space and indentation should be as per style guide

File: app/views/invoices/_form.html.haml [hunk 11]
```diff
@@ -82,11 +97,27 @@
...
+        if(!isNaN(parseFloat($("#invoice_total").val()))) {
 
-  $(function() {
-    $('.datepicker').datepicker();
+            $($("#invoice_discount").val("0.00"));
```
Comment (amitpatelx) [inline app/views/invoices/_form.html.haml]: What this statement does?

Comment (PragatiDoshi-BTC) [inline app/views/invoices/_form.html.haml]: on change of total invoice value, invoice discount will be 0.00.

**Comment by ParthivPatel-BTC** [comment]
CSS classes are not applying properly on success/error messages. Please test whole application properly and fix it.

**Pull request merged**